File: src/my_dog_node/my_dog_node/trot.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64MultiArray
import math
import time

class TrotGait(Node):
    def __init__(self):
        super().__init__('trot_gait_node')
        
        # Паблишер для отправки команд контроллеру
        self.publisher_ = self.create_publisher(
            Float64MultiArray, 
            '/joint_group_position_controller/commands', 
            10
        )
        
        # Таймер: 50 Гц (каждые 0.02 сек)
        self.timer = self.create_timer(0.02, self.timer_callback)
        
        # === РЕАЛЬНЫЕ РАЗМЕРЫ РОБОТА (Метры) ===
        self.L1 = 0.06    # Hip (вбок)
        self.L2 = 0.144   # Thigh (бедро)
        self.L3 = 0.1525  # Shin (голень)
        
        # === ПАРАМЕТРЫ ПОХОДКИ ===
        self.walking_height = -0.16  # Присядем еще ниже (стабильнее)
        self.step_length = 0.04      # Короткий шаг (4 см)
        self.step_height = 0.03      # Низкий подъем
        self.period = 2.0            # ОЧЕНЬ медленный цикл (2 сек)
        
        self.start_time = time.time()
        self.get_logger().info("Trot Gait Started! Watch out!")

    def inverse_kinematics(self, x, y, z):
        """
        Считает углы для одной лапы.
        x, y, z - координаты стопы относительно крепления ноги (плеча).
        """
        # 1. Hip (Roll) - пока упростим и оставим 0 (без движения вбок)
        theta1 = 0.0
        
        # 2. Thigh & Shin (Pitch)
        # Учитываем, что L1 сдвигает ногу вбок, но IK решаем в плоскости ноги
        # Проекция на плоскость вращения бедра:
        dist_2d = math.sqrt(x**2 + z**2)
        
        # Защита от выхода за пределы длины ноги
        max_reach = self.L2 + self.L3 - 0.005
        if dist_2d > max_reach:
            dist_2d = max_reach

        # Теорема косинусов для колена (Shin)
        # c^2 = a^2 + b^2 - 2ab cos(gamma)
        cos_knee = (self.L2**2 + self.L3**2 - dist_2d**2) / (2 * self.L2 * self.L3)
        cos_knee = max(-1.0, min(1.0, cos_knee)) # Clamp
        phi = math.acos(cos_knee)
        
        # Угол голени (сгиб назад)
        theta3 = -(math.pi - phi)

        # Теорема косинусов для бедра (Thigh)
        # Угол наклона вектора к горизонту
        alpha = math.atan2(x, -z) 
        
        # Внутренний угол треугольника
        cos_beta = (self.L2**2 + dist_2d**2 - self.L3**2) / (2 * self.L2 * dist_2d)
        cos_beta = max(-1.0, min(1.0, cos_beta))
        beta = math.acos(cos_beta)
        
        theta2 = alpha + beta
        
        return theta1, theta2, theta3

    def get_leg_trajectory(self, t, phase_offset):
        """
        Вычисляет координаты (x, z) для лапы в текущий момент времени.
        t - текущее время
        phase_offset - смещение фазы (0 для одной диагонали, 0.5 для другой)
        """
        # Приводим время к циклу [0, 1)
        cycle_t = (t / self.period + phase_offset) % 1.0
        
        # Фаза переноса (Swing) - 0.0 до 0.5
        if cycle_t < 0.5:
            # Прогресс внутри фазы переноса (от 0 до 1)
            swing_progress = cycle_t / 0.5
            
            # X: Двигаемся вперед от -step/2 до +step/2
            # Используем косинус для плавного разгона и торможения
            x = self.step_length/2 * math.cos(math.pi * swing_progress)
            
            # Z: Поднимаем лапу по синусоиде (полукруг)
            z = self.walking_height + self.step_height * math.sin(math.pi * swing_progress)
            
        # Фаза опоры (Stance) - 0.5 до 1.0
        else:
            # Прогресс внутри фазы опоры
            stance_progress = (cycle_t - 0.5) / 0.5
            
            # X: Толкаем землю НАЗАД (от плюса к минусу)
            # Оригинальная формула была (cos), мы её инвертируем
            # Было: raw_x = self.step_length/2 * math.cos(...)
            # Стало: добавляем минус перед math.cos
            x = self.step_length/2 * -math.cos(math.pi * stance_progress)
            
            z = self.walking_height

        return x, 0.0, z # Y пока 0

    def timer_callback(self):
        # Текущее время
        t = time.time() - self.start_time
        
        # Генерируем траектории для двух диагоналей
        # Диагональ 1: FL и RR (offset 0.0)
        x1, y1, z1 = self.get_leg_trajectory(t, 0.0)
        
        # Диагональ 2: FR и RL (offset 0.5 - в противофазе)
        x2, y2, z2 = self.get_leg_trajectory(t, 0.5)

        # Считаем IK для каждой лапы
        # Порядок: FL, FR, RL, RR
        
        # Front Left (Diag 1)
        q_fl = self.inverse_kinematics(x1, y1, z1)
        
        # Front Right (Diag 2) -> Для правой стороны hip (q1) инвертируем, но он пока 0
        q_fr = self.inverse_kinematics(x2, y2, z2)
        
        # Rear Left (Diag 2)
        q_rl = self.inverse_kinematics(x2, y2, z2)
        
        # Rear Right (Diag 1)
        q_rr = self.inverse_kinematics(x1, y1, z1)

        # Собираем сообщение (12 углов)
        # Структура: [Hip, Thigh, Shin] для каждой лапы
        # ВАЖНО: Для правых лап (FR, RR) часто нужно инвертировать Hip, 
        # но так как Hip=0, это не важно.
        
        msg = Float64MultiArray()
        msg.data = [
            q_fl[0], q_fl[1], q_fl[2],   # FL
            q_fr[0], q_fr[1], q_fr[2],   # FR
            q_rl[0], q_rl[1], q_rl[2],   # RL
            q_rr[0], q_rr[1], q_rr[2]    # RR
        ]
        
        self.publisher_.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    node = TrotGait()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/dog_brain/src/trot.cpp
#include <chrono>
#include <cmath>
#include <memory>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"
#include "geometry_msgs/msg/twist.hpp" 

using namespace std::chrono_literals;

class TrotNode : public rclcpp::Node
{
public:
    TrotNode() : Node("trot_node")
    {
        publisher_ = this->create_publisher<std_msgs::msg::Float64MultiArray>(
            "/joint_group_position_controller/commands", 10);
            
        timer_ = this->create_wall_timer(
            20ms, std::bind(&TrotNode::timer_callback, this));
            
        start_time_ = this->now();
        
        subscription_ = this->create_subscription<geometry_msgs::msg::Twist>(
            "/cmd_vel", 10, std::bind(&TrotNode::cmd_vel_callback, this, std::placeholders::_1));
        
        RCLCPP_INFO(this->get_logger(), "Ready for Teleop! Use /cmd_vel");
    }

private:
    const double L1 = 0.06;
    const double L2 = 0.144;
    const double L3 = 0.1525;
    
    // Целевые скорости
    double target_linear_x_ = 0.0;
    double target_angular_z_ = 0.0;

    // Подписчик
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr subscription_;

    // === ПАРАМЕТРЫ ===
    // === НАСТРОЙКИ "ТАНК" ===
    const double WALKING_HEIGHT = -0.22; // Ползаем на брюхе (макс стабильность)
    const double STEP_LENGTH = 0.02;     // 2 сантиметра! Почти на месте.
    const double STEP_HEIGHT = 0.04;
    const double PERIOD = 1.0;
    const double X_OFFSET = 0.0;         // Без хитростей

    struct Point { double x; double y; double z; };

    std::vector<double> inverse_kinematics(double x, double y, double z)
    {
        double target_x = x + X_OFFSET;
        double theta1 = 0.0;
        
        double abs_z = std::abs(z);
        double dist_2d = std::sqrt(target_x * target_x + abs_z * abs_z);
        
        double max_reach = L2 + L3 - 0.005;
        if (dist_2d > max_reach) dist_2d = max_reach;

        double cos_knee = (L2 * L2 + L3 * L3 - dist_2d * dist_2d) / (2 * L2 * L3);
        if (cos_knee > 1.0) cos_knee = 1.0;
        if (cos_knee < -1.0) cos_knee = -1.0;
        double phi = std::acos(cos_knee);
        double theta3 = -(M_PI - phi); 

        double alpha = std::atan2(target_x, abs_z); 
        double cos_beta = (L2 * L2 + dist_2d * dist_2d - L3 * L3) / (2 * L2 * dist_2d);
        if (cos_beta > 1.0) cos_beta = 1.0;
        if (cos_beta < -1.0) cos_beta = -1.0;
        double beta = std::acos(cos_beta);
        double theta2 = alpha + beta; 

        return {theta1, theta2, theta3};
    }

    void cmd_vel_callback(const geometry_msgs::msg::Twist::SharedPtr msg){
        target_linear_x_ = msg->linear.x;
        target_angular_z_ = msg->angular.z;
    }

    Point get_leg_trajectory(double t, double phase_offset)
    {
        double cycle_t = std::fmod((t / PERIOD + phase_offset), 1.0);
        double x = 0.0;
        double z = WALKING_HEIGHT;
        
        // Масштабируем длину шага от скорости
        double current_step_length = STEP_LENGTH * target_linear_x_; 

       if (std::abs(current_step_length) < 0.001) {
         // СТОИМ НА МЕСТЕ
         // Вместо просто высоты, давай опустим его на 1 см ниже.
         // Это создаст сильное давление на стопы и остановит дрейф.
         return {0.0, 0.0, WALKING_HEIGHT - 0.01}; 
    }
        if (cycle_t < 0.5)
        {
            // === SWING (Перенос) ===
            double swing_progress = cycle_t / 0.5; // <--- ВОТ ЭТОЙ СТРОКИ НЕ БЫЛО
            
            // X: Двигаемся ВПЕРЕД (-cos идет от -1 к 1)
            x = (current_step_length / 2.0) * -std::cos(M_PI * swing_progress);
            
            z = WALKING_HEIGHT + STEP_HEIGHT * std::sin(M_PI * swing_progress);
        }
        
        else
        {
            // === STANCE (Опора) ===
            double stance_progress = (cycle_t - 0.5) / 0.5; // <--- И ЭТОЙ НЕ БЫЛО
            
            // X: Толкаем землю НАЗАД (линейно от 1 к -1)
            x = (current_step_length / 2.0) * (1.0 - 2.0 * stance_progress); 
            z = WALKING_HEIGHT;
        }
        
        return {x, 0.0, z};
    }
    void timer_callback()
    {
        double t = (this->now() - start_time_).seconds();
        
        Point p1 = get_leg_trajectory(t, 0.0);
        Point p2 = get_leg_trajectory(t, 0.5);

        auto fl = inverse_kinematics(p1.x, 0.0, p1.z);
        auto fr = inverse_kinematics(p2.x, 0.0, p2.z);
        auto rl = inverse_kinematics(p2.x, 0.0, p2.z);
        auto rr = inverse_kinematics(p1.x, 0.0, p1.z);

        auto msg = std_msgs::msg::Float64MultiArray();
        msg.data = {
            fl[0], fl[1], fl[2],
            fr[0], fr[1], fr[2],
            rl[0], rl[1], rl[2],
            rr[0], rr[1], rr[2]
        };
        publisher_->publish(msg);
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr publisher_;
    rclcpp::Time start_time_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TrotNode>());
    rclcpp::shutdown();
    return 0;
}